protocols/kad/src/record/store/memory.rs

```diff
@@ -131,8 +131,10 @@ impl<'a> RecordStore<'a> for MemoryStore {
         Ok(())
     }
 
-    fn remove(&'a mut self, k: &Key) {
+    fn remove(&'a mut self, k: &Key) -> Result<()> {
```

Comment:
Can you explain in more detail, what benefits you are expecting to see from this?

Removing an element should be an idempotent operation in my opinion. What failure scenarios would you like to express here?

Reply:
The current `RecordStore` implies the `MemoryStore` implementation that could be considered infallible which is not the case for most other custom implementations. Custom stores implemented with DB or file layers are always fallible for multiple reasons.

Just an example of a possible error for the `remove` operation: "No write permissions for the database file". 

Also, the `remove` operation is just an example of the whole `RecordStore` API. Specifically, the most problems I have are with the `records` operation. My database returns a `Result` for such iterations and I needed to create a special empty iterator with some error logging which is not a good practice IMHO.

The original issue contains a similar concern from other users: https://github.com/libp2p/rust-libp2p/issues/3035#issuecomment-1283866786

Reply:
I understand that DB or file system APIs are fallible. But what is the `NetworkBehaviour` supposed to do with the error? Log it? If there is no reasonable error handling strategy, we might as well not allow the implementation to return one.

Reply:
An implementation of `RecordStore` that interacts with a DB could also just represent a `Handle` that sends commands to the actual implementation, similar to how SQL DBs usually write to a journal first before modifying data on disk.

Reply:
> I understand that DB or file system APIs are fallible. But what is the `NetworkBehaviour` supposed to do with the error? Log it? If there is no reasonable error handling strategy, we might as well not allow the implementation to return one.

That depends on the use case and libp2p implementation (architecture). Some use cases are compatible with the `Result` already, for example, `remove_record` from `Behaviour` will return it to the caller. Some with the current code are hard to work with and logging seems the easiest change. I agree, that it will add complexity to the design, however, IMHO it makes the API more idiomatic and reflects a broader set of use cases. 

> An implementation of RecordStore that interacts with a DB could also just represent a Handle that sends commands to the actual implementation, similar to how SQL DBs usually write to a journal first before modifying data on disk.

I'm not sure I follow here. There are a lot of ways to implement the storage indeed and some will hide the errors.

Reply:
> > An implementation of RecordStore that interacts with a DB could also just represent a Handle that sends commands to the actual implementation, similar to how SQL DBs usually write to a journal first before modifying data on disk.
> 
> I'm not sure I follow here. There are a lot of ways to implement the storage indeed and some will hide the errors.

Sorry for me brevity. I'll try to expand:

In general, removing an entry from a set is an idempotent operation: If it is present, remove it. If it is not present, do nothing.
IO-failures are typically ephemeral. Permission errors can silently drop the "request" and not remove the element.

This means that you could implement your `RecordStore` as a channel where calling `RecordStore::remove` doesn't actually remove the record but sends an (asynchronous) `Remove` message to the actual implementation. Same thing for insert, etc.

As long as you retain the order of the messages, your actual store of records can apply those messages and should end up in the same state as if the `RecordStore` functions would have synchronously modified the data. The difference there however is that appending one of these messages is a non-fallible operation. It is then the responsibility of whoever reads these messages to apply them correctly.

I am not convinced it makes sense to make these functions fallible, given that they are called in various places where we cannot apply a useful error handling strategy.

Reply:
This design could be implemented and has useful features (e.g.: you mentioned "handling ephemeral errors") but it also has its limitations - what if the error persists ("Disk is full.")? In this case, if we use an unbounded buffer - we will get memory overflow and if we have a limited buffer we will get the same situation with error handling  (or we will need to delete some requests). 

I understand that you see the architecture as optimized for infallible usage. My initial request was because I encountered an issue with the implementation when wrapping the DB as storage. 
Probably, if we could ask @rklaehn about his experience https://github.com/libp2p/rust-libp2p/issues/3035#issuecomment-1283866786 we could add additional arguments to our discussion.

Reply:
> but it also has its limitations - what if the error persists ("Disk is full.")?

Yes that is a fair point.

I see two solutions:

1. The entire `RecordStore` could live outside of `rust-libp2p` and the `NetworkBehaviour` emits an event per incoming request that needs to be answered.
2. We provide a fallible, and potentially async `RecordStore` interface which we directly integrate into the `ConnectionHandler`. This would be similar to the ideas in #2852. If we integrate the `RecordStore` directly into the `ConnectionHandler`, having the functions fallible makes a lot more sense because we could just fail the entire stream and return an error back directly in the context of the other party's request. We are also in an `async` context there so the implementation of `RecordStore` could even query a database or other things.

(2) is quite a paradigm shift but I think it would enable a lot of usecases and cut down the codesize of `libp2p-kad` a lot.

Reply:
One important property I’d like to uphold is that NetworkBehaviour is about networking stuff, connection handlers care about handling connections, and other things live outside these abstractions — at least in terms of code module structure. Kademlia might require an implementation of RecordStore in its constructor, and it might use the provided functions from the NetworkBehaviour or from the ConnectionHandler, but the actual RecordStore should live outside these two pieces — also regarding resource usage.

One somewhat similar case is bitswap, where our store also is structured like this, but it has a fully synchronous API and is spawned inside a dedicated thread, with communication over async channels. Failures will need to be bubbled up to the network peer, plus logging (with careful tuning of levels — only things like “disk full” warrant a WARN or ERROR, everything else should stay below INFO, or preferably below DEBUG for per-interaction events).

Reply:
I would vote for option 2 (direct integration into the ConnectionHandler). However, it seems like a major change. So, my question is: whether there is any benefit from this PR as the preparation for that refactoring (that unlikely will make it in the nearest working plan)? Otherwise, feel free to reject it.

Reply:
> Kademlia might require an implementation of RecordStore in its constructor
Did you mean this one? https://github.com/libp2p/rust-libp2p/blob/5b4eab7bafe6cba8fea6c806027ea680662b7cc6/protocols/kad/src/behaviour.rs#L405

Reply:
> One important property I’d like to uphold is that NetworkBehaviour is about networking stuff, connection handlers care about handling connections, and other things live outside these abstractions — at least in terms of code module structure. Kademlia might require an implementation of RecordStore in its constructor, and it might use the provided functions from the NetworkBehaviour or from the ConnectionHandler, but the actual RecordStore should live outside these two pieces — also regarding resource usage.

The reason I am challenging it is because we pay a very big price for this separation and the gains are hard to quantify for me:

- There is a lot of message passing between `ConnectionHandler`s and `NetworkBehaviour`s, only because we can't access a shared resource directly from the handlers.
- Because of the message passing, we can't use async-await to describe our protocols, despite them being mostly request-response.

The interface to `RecordStore` could still look something like this:

```rust
trait RecordStore: Clone {
	fn get_value(&self, key: Key) -> GetValueFuture;
}
```

This would force implementations to use channels or other things to communicate with the centralised store (self receiver is not mutable and future doesn't have a lifetime). Thus, we would still uphold the "share memory by communicating" mantra and the connections themselves would only wait for a channel wake-up but not perform any other work.

At the same time though, this `RecordStore` can be referenced within each `ConnectionHandler`, the protocol can be expressed in a few LoC and even adding things like timeouts for retrieving something from the store can trivially be added.